source/simulator/assemblers/advection.cc

```diff
@@ -494,7 +494,11 @@ namespace aspect
               ExcInternalError());
       const bool cell_has_periodic_neighbor = cell->has_periodic_neighbor (face_no);
 
-      if (!(face->has_children()))
+      if ((!face->at_boundary() && !face->has_children())
+          ||
+          (face->at_boundary() && cell->periodic_neighbor_is_coarser(face_no))
+          ||
+          (face->at_boundary() && neighbor->level () == cell->level () && neighbor->active()))
```

Comment:
This condition is difficult to read. Start with what you want to do, namely exclude the case of a face whose neighbor is more refined. Then you could write instead
```
  if ( ! neighbor->has_children())
```
This works because `neighbor` (== `cell->neighbor_or_periodic_neighbor (face_no)`) will always be either (i) a cell at a coarser level if there is no neighbor at the same level, or (ii) a cell at the same level, regardless of whether that neighbor has children or not.